x/subscription/keeper/subscription.go

```diff
@@ -256,20 +188,91 @@ func (k Keeper) CreateSubscription(
 		)
 	}
 
-	k.SetSubscription(ctx, sub)
+	if !found {
+		expiry := NextMonth(ctx.BlockTime()).UTC().Unix()
+		k.subsTS.AddTimerByBlockTime(ctx, uint64(expiry), []byte(consumer), []byte{})
+		err = k.subsFS.AppendEntry(ctx, consumer, block, &sub)
+	} else {
+		k.subsFS.ModifyEntry(ctx, consumer, sub.Block, &sub)
+	}
 
-	return nil
+	return err
+}
+
+func (k Keeper) advanceMonth(ctx sdk.Context, subkey []byte) {
+	date := ctx.BlockTime()
+	block := uint64(ctx.BlockHeight())
+	consumer := string(subkey)
+
+	var sub types.Subscription
+	_ = k.subsFS.FindEntry(ctx, consumer, block, &sub)
+
+	if sub.DurationLeft == 0 {
+		utils.LavaFormatError("critical: negative duration for subscription, extend by 1 month",
+			fmt.Errorf("negative duration in AdvanceMonth()"),
+			utils.Attribute{Key: "consumer", Value: consumer},
+			utils.Attribute{Key: "plan", Value: sub.PlanIndex},
+			utils.Attribute{Key: "block", Value: block},
+		)
+		// normally would panic! but can "recover" by auto-extending by 1 month
+		expiry := NextMonth(date).UTC().Unix()
```

Comment:
i this case we should delete the subscription? or do nothing
adding a month wont really help

Reply:
i agree with the error but i dont thing adding a month is doing nothing, especially if the subscription has reached 0, this means that the problem will happen next month again? why not just delete it and move on?

Reply:
Something went wrong, I prefer to leave things they are for forensics and recovery.
There is no harm in moving the expiry one month forward - no additional CUs.